File: app/src/main/java/com/huwei/sweetmusicplayer/business/ArtistInfoActivity.java
package com.huwei.sweetmusicplayer.business;


import android.content.Context;
import android.content.Intent;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;


import com.google.gson.Gson;
import com.huwei.sweetmusicplayer.R;
import com.huwei.sweetmusicplayer.SweetApplication;
import com.huwei.sweetmusicplayer.business.baidumusic.po.ArtistInfo;
import com.huwei.sweetmusicplayer.business.fragments.AlbumListFragment_;
import com.huwei.sweetmusicplayer.business.fragments.BaseScrollTabFragment;
import com.huwei.sweetmusicplayer.business.fragments.SongListFragment_;
import com.huwei.sweetmusicplayer.business.interfaces.IAdjustListView;
import com.huwei.sweetmusicplayer.business.interfaces.IListViewScroll;
import com.huwei.sweetmusicplayer.business.ui.adapters.PagerAdapter;
import com.huwei.sweetmusicplayer.business.ui.views.ArtistInfoView;
import com.huwei.sweetmusicplayer.business.ui.widgets.GradientToolbar;
import com.huwei.sweetmusicplayer.util.BaiduMusicUtil;
import com.huwei.sweetmusicplayer.util.HttpHandler;
import com.nineoldandroids.view.ViewHelper;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.PageSelected;
import org.androidannotations.annotations.ViewById;

import java.util.Arrays;

/**
 * 在线歌手信息
 *
 * @author jerry
 * @date 2015-12-22
 */
@EActivity(R.layout.activity_artist_info)
public class ArtistInfoActivity extends BaseActivity implements IListViewScroll {

    public static final String ARTISTID = "artist_id";
    public static final String TINGUID = "tinguid";

    public static final String RES_TITLE[] = new String[]{"歌曲", "专辑"};

    @ViewById
    GradientToolbar gtoolbar;
    @ViewById(R.id.toolbar)
    Toolbar toolbar;
    @ViewById(R.id.view_artist_info)
    ArtistInfoView mArtistInfoView;
    @ViewById(R.id.viewpager)
    public ViewPager mViewPager;
    @ViewById(R.id.tabs)
    TabLayout mTab;
    @ViewById(R.id.ll_flow_header)
    View mHeaderFlow;

    private String ting_uid;
    private String artist_id;

    private BaseScrollTabFragment mSongListFragment;
    private BaseScrollTabFragment mAlbumListFragment;
    private PagerAdapter mPagerAdapter;

    private int mLimitHeight; //headerflow 移动的最小限制

    public static Intent getStartActIntent(Context from, String ting_uid, String artist_id) {
        Intent intent = new Intent(from, ArtistInfoActivity_.class);
        intent.putExtra(TINGUID, ting_uid);
        intent.putExtra(ARTISTID, artist_id);
        return intent;
    }

    @AfterViews
    void init() {
        ting_uid = getIntent().getStringExtra(TINGUID);
        artist_id = getIntent().getStringExtra(ARTISTID);

        initToolBar();
        initMeasure();
        initViewPager();
        intTab();
        initGToolBar();
        getArtistInfo();
    }

    void initToolBar() {
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);

        toolbar.setTitle(R.string.activity_artist_info);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackClicked(v);
            }
        });
    }

    void initMeasure() {
        mHeaderFlow.measure(View.MeasureSpec.makeMeasureSpec(SweetApplication.mScreenWidth, View.MeasureSpec.EXACTLY), View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
        gtoolbar.measure(View.MeasureSpec.makeMeasureSpec(SweetApplication.mScreenWidth, View.MeasureSpec.EXACTLY), View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
        mLimitHeight = mArtistInfoView.getMeasuredHeight() - gtoolbar.getMeasuredHeight();
        Log.i(TAG, "mArtistInfoView:" + mArtistInfoView.getMeasuredHeight() + " gtoolbar.getMeasuredHeight():" + gtoolbar.getMeasuredHeight() + "   mLimitHeight:" + mLimitHeight);
    }

    void initViewPager() {
        mSongListFragment = SongListFragment_.builder().artist_id(artist_id).ting_uid(ting_uid).build();
        mAlbumListFragment = AlbumListFragment_.builder().artist_id(artist_id).ting_uid(ting_uid).build();
        mPagerAdapter = new PagerAdapter(getSupportFragmentManager(), Arrays.asList((Fragment) mSongListFragment, mAlbumListFragment)) {
            @Override
            public CharSequence getPageTitle(int position) {
                return RES_TITLE[position];
            }
        };
        mViewPager.setAdapter(mPagerAdapter);

        for (int i = 0; i < mPagerAdapter.getCount(); i++) {
            BaseScrollTabFragment fragment = (BaseScrollTabFragment) mPagerAdapter.getItem(i);
            fragment.setIListViewScroll(this);
            fragment.setFlowHeight(mHeaderFlow.getMeasuredHeight());
            fragment.setlowLimitHeight(mLimitHeight);
        }
    }

    void intTab() {
        mTab.setupWithViewPager(mViewPager);
    }

    void initGToolBar() {
        gtoolbar.bindHeaderView(mArtistInfoView);
    }


    void getArtistInfo() {
        BaiduMusicUtil.getArtistInfo(ting_uid, artist_id, new HttpHandler() {
            @Override
            public void onSuccess(String response) {
                ArtistInfo artistInfo = new Gson().fromJson(response, ArtistInfo.class);
                mArtistInfoView.bind(artistInfo, gtoolbar);
                TabLayout.Tab mTabSong = mTab.getTabAt(0);
                if (mTabSong != null) {
                    mTabSong.setText("歌曲(" + artistInfo.songs_total + ")");
                }
                TabLayout.Tab mTabAlbum = mTab.getTabAt(1);
                if (mTabAlbum != null) {
                    mTabAlbum.setText("专辑(" + artistInfo.albums_total + ")");
                }

                if (artistInfo != null) {
                    gtoolbar.setGradientTitle(artistInfo.name);
                }
            }
        });
    }

    @PageSelected(R.id.viewpager)
    void onPageSelected(int position) {
        // Something Here
        Log.i(TAG, "viewPager position:" + position);

        IAdjustListView iAdjustListView = (IAdjustListView) mPagerAdapter.getItem(position);
        iAdjustListView.adjustListView(getOffestY());
    }

    @Override
    public void scrollY(int scrollY) {
        float translationY = mHeaderFlow.getTranslationY();
        Log.i(TAG, "-scrollY:" + scrollY + "    translationY:" + translationY);
        if (translationY != -scrollY) {
            ViewHelper.setTranslationY(mHeaderFlow, -scrollY);
        }
        gtoolbar.adjustHeaderViewAndTitle();
    }

    public int getOffestY() {
        int paddingHeight = (int) (mHeaderFlow.getTranslationY() + mHeaderFlow.getMeasuredHeight());
        Log.i(TAG, "paddingHeight:" + paddingHeight + "mHeaderFlow.getTranslationY():" + mHeaderFlow.getTranslationY());
        return paddingHeight;
    }
}

File: app/src/main/java/com/huwei/sweetmusicplayer/business/ArtistInfoActivity.kt
package com.huwei.sweetmusicplayer.business

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.ViewCompat.setTranslationY
import android.support.v4.view.ViewPager
import android.util.Log
import android.view.View

import com.huwei.sweetmusicplayer.R
import com.huwei.sweetmusicplayer.SweetApplication
import com.huwei.sweetmusicplayer.data.models.baidumusic.po.ArtistInfo
import com.huwei.sweetmusicplayer.business.onlinesearch.AlbumListFragment
import com.huwei.sweetmusicplayer.business.onlinesearch.SongListFragment
import com.huwei.sweetmusicplayer.business.interfaces.IAdjustListView
import com.huwei.sweetmusicplayer.business.interfaces.IListViewScroll
import com.huwei.sweetmusicplayer.data.api.RetrofitFactory
import com.huwei.sweetmusicplayer.data.api.SimpleObserver
import com.huwei.sweetmusicplayer.data.api.baidu.BaiduMusicService
import com.huwei.sweetmusicplayer.ui.adapters.PagerAdapter
import com.huwei.sweetmusicplayer.frameworks.BundleBuilder
import com.huwei.sweetmusicplayer.util.Utils

import java.util.Arrays

import com.huwei.sweetmusicplayer.data.contants.IntentExtra.EXTRA_ARTIST_ID
import com.huwei.sweetmusicplayer.data.contants.IntentExtra.EXTRA_TING_UID
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_artist_info.*
import kotlinx.android.synthetic.main.layout_gradient_toolbar.*

/**
 * 在线歌手信息
 *
 * @author jerry
 * @date 2015-12-22
 */
class ArtistInfoActivity : BottomPlayActivity(), IListViewScroll {

    private var ting_uid: String = ""
    private var artist_id: String = ""

    private var mSongListFragment: BaseScrollTabFragment? = null
    private var mAlbumListFragment: BaseScrollTabFragment? = null
    private var mPagerAdapter: PagerAdapter? = null

    private var mLimitHeight: Int = 0 //headerflow 移动的最小限制

    val offestY: Int
        get() {
            val paddingHeight = (ll_flow_header!!.translationY + ll_flow_header!!.measuredHeight).toInt()
            Log.i(TAG, "paddingHeight:" + paddingHeight + "ll_flow_header.getTranslationY():" + ll_flow_header!!.translationY)
            return paddingHeight
        }

    override fun isNeedStatusView(): Boolean {
        return false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_artist_info)

        ting_uid = intent.getStringExtra(TINGUID)
        artist_id = intent.getStringExtra(ARTISTID)

        initToolBar()
        initMeasure()
        initViewPager()
        intTab()
        initGToolBar()
        getArtistInfo()
    }

    internal fun initToolBar() {
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setHomeButtonEnabled(true)

        toolbar!!.setTitle(R.string.activity_artist_info)
        toolbar!!.setNavigationOnClickListener { v -> onBackClicked(v) }
    }

    internal fun initMeasure() {
        ll_flow_header!!.measure(View.MeasureSpec.makeMeasureSpec(SweetApplication.mScreenWidth, View.MeasureSpec.EXACTLY), View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED))
        gtoolbar!!.measure(View.MeasureSpec.makeMeasureSpec(SweetApplication.mScreenWidth, View.MeasureSpec.EXACTLY), View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED))
        //由于这里gtoolbar算出来只是56dp,需要加上状态栏的高度
        mLimitHeight = view_artist_info!!.measuredHeight - (gtoolbar!!.measuredHeight + Utils.getStatusBarHeight())
        Log.i(TAG, "view_artist_info:" + view_artist_info!!.measuredHeight + " gtoolbar.getMeasuredHeight():" + gtoolbar!!.measuredHeight + "   mLimitHeight:" + mLimitHeight)
    }

    internal fun initViewPager() {
        mSongListFragment = SongListFragment()
        mSongListFragment!!.arguments = BundleBuilder().extra(EXTRA_ARTIST_ID, artist_id!!).extra(EXTRA_TING_UID, ting_uid!!).build()
        mAlbumListFragment = AlbumListFragment()
        mAlbumListFragment!!.arguments = BundleBuilder().extra(EXTRA_ARTIST_ID, artist_id!!).extra(EXTRA_TING_UID, ting_uid!!).build()
        mPagerAdapter = object : PagerAdapter(supportFragmentManager, Arrays.asList<Fragment>(mSongListFragment as Fragment?, mAlbumListFragment)) {
            override fun getPageTitle(position: Int): CharSequence {
                return RES_TITLE[position]
            }
        }
        viewpager!!.adapter = mPagerAdapter
        viewpager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {
            }

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            }

            override fun onPageSelected(position: Int) {
                // Something Here
                Log.i(TAG, "viewPager position:" + position)

                val iAdjustListView = mPagerAdapter!!.getItem(position) as IAdjustListView
                iAdjustListView.adjustListView(offestY)
            }
        })

        for (i in 0 until mPagerAdapter!!.count) {
            val fragment = mPagerAdapter!!.getItem(i) as BaseScrollTabFragment
            fragment.setIListViewScroll(this)
            fragment.setFlowHeight(ll_flow_header!!.measuredHeight)
            fragment.setlowLimitHeight(mLimitHeight)
        }
    }

    internal fun intTab() {
        tabLayout!!.setupWithViewPager(viewpager)
    }

    internal fun initGToolBar() {
        gtoolbar!!.setTitle(R.string.activity_artist_info)
        gtoolbar!!.bindHeaderView(view_artist_info)
    }


    internal fun getArtistInfo() {
        RetrofitFactory.create(BaiduMusicService::class.java)
                .getArtistInfo(ting_uid, artist_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SimpleObserver<ArtistInfo>() {
                    override fun onSuccess(resp: ArtistInfo) {
                        view_artist_info!!.bind(resp!!, gtoolbar)
                        val mTabSong = tabLayout!!.getTabAt(0)
                        if (mTabSong != null) {
                            mTabSong.text = "歌曲(" + resp.songs_total + ")"
                        }
                        val mTabAlbum = tabLayout!!.getTabAt(1)
                        if (mTabAlbum != null) {
                            mTabAlbum.text = "专辑(" + resp.albums_total + ")"
                        }

                        if (resp != null) {
                            gtoolbar!!.setGradientTitle(resp.name)
                        }
                    }
                })
    }

    override fun scrollY(scrollY: Int) {
        val translationY = ll_flow_header!!.translationY
        Log.i(TAG, "-scrollY:$scrollY    translationY:$translationY")
        if (translationY != (-scrollY).toFloat()) {
            setTranslationY(ll_flow_header!!, (-scrollY).toFloat())
        }
        gtoolbar!!.adjustHeaderViewAndTitle()
    }

    companion object {

        val ARTISTID = "artist_id"
        val TINGUID = "tinguid"

        val RES_TITLE = arrayOf("歌曲", "专辑")

        fun getStartActIntent(from: Context, ting_uid: String, artist_id: String): Intent {
            val intent = Intent(from, ArtistInfoActivity::class.java)
            intent.putExtra(TINGUID, ting_uid)
            intent.putExtra(ARTISTID, artist_id)
            return intent
        }
    }
}
